### ISSUE TITLE
Metadata: If several authors, only one is displayed

### ISSUE THREAD

**Issue by gerhaher** [koreader/koreader#3583]
* KOReader version: v2015.11-1432-ga692b5a
* Device: KA1

#### Issue
If a book have several authors and if you have separated each name with '&', only the first name will be displayed. This occurs for epubs, not for pdfs.

#### Steps to reproduce

1. Transfer a epub book to your e-reader with author name 'A & B'. (And if you want, for comparison, a pdf with the same names)
2. Choose, for example, "Detailed list with metadata, no images" as display mode
3. The result will be 'A' for the epub and 'A & B' for the pdf

**Comment by poire-z** [comment]
It works for me if the '&' is properly encoded as `&amp;` or `&#38;` (so it should if you or the software you used to edit the metadata have properly encoded it).

If in OEBPS/content.opf (or similar), i have:
```xml
  <metadata>
    <dc:title>Terre</dc:title>
    <dc:creator>Wikipedia &copy; FR &amp; me &#38; myself & mom</dc:creator>
  </metadata>
```
(as you see, github highlights the standalone '&' as it's not proper XML, and I guess this would make the EPUB fails some validation tools).

it displays as:
<kbd>![metaamp](https://user-images.githubusercontent.com/24273478/34717847-be798980-f534-11e7-8d55-bedabb9491f2.png)</kbd>

**Comment by gerhaher** [comment]
Thanks for your answer.

I checked the content.opf for some books. Here are two examples how it looks for me:

<dc:creator id="id">Greg King</dc:creator>
<dc:creator id="id-1">Penny Wilson</dc:creator>

<dc:creator xmlns:ns1="http://www.idpf.org/2007/opf" ns1:role="aut" ns1:file-as="Berglund, Tobias &amp; Sennerteg, Niclas">Tobias Berglund</dc:creator>
<dc:creator xmlns:ns0="http://www.idpf.org/2007/opf" ns0:role="aut">Niclas Sennerteg</dc:creator>

(Koreader displays only Greg King and Tobias Berglund.)


I use Calibre. So the problem is somewhere in the settings in Calibre? 
None of my epubs show more than one author.

**Comment by poire-z** [comment]
Oh, ok, you have multiple `<dc:creator>` elements, and it looks like the specs allow for that:
http://www.idpf.org/epub/20/spec/OPF_2.0.1_draft.htm#Section2.2.2

There are multiple discussions about that and limits of reading systems:
http://idpf.org/forum/topic-943
http://idpf.org/forum/topic-1238
https://github.com/w3c/publ-epub-revision/issues/369

crengine (our epub renderer) uses only the first one.
We could easily give it the same treatment we gave it for `<db:subject>` (multiple keywords, made as a single string with them separated by `; ` ) if that's a really needed feature
https://github.com/koreader/crengine/blob/a3eeb1bf2663506fa84c38470aca6fa66d9636df/crengine/src/epubfmt.cpp#L769-L794

May be calibre has a settings to merge multiple authors into a single `<dc:creator>` element?

**Comment by poire-z** [comment]
Should we do something about multiple instances of these metadata elements?

#### creator:
From the specs:
_Publications with multiple co-authors should provide `multiple creator` elements, each containing one author. The order of creator elements is presumed to define the order in which the creators' names should be presented by the Reading System.
This specification recommends that the content of the creator elements hold the text for a single name as it would be presented to the Reader._ (me: what does that mean ?) 

If we should concatenate the multiple `<dc:creator>` into our single `authors` field, what's should the separator be? `; ` like for keywords? (I guess `,` could look strange, as individual authors could formatted as `Lennon, John` and `Marley, Bob`, and we don't want to have Authors: `Lennon, John, Marley, Bob`.
It may also become large and render the display in very tiny fonts so it can be displayed fully, in the various coverbrowser plugin modes (with the solution to "Ignore these metadata" in the Hold menu to get back to filename)

#### title:
_The title of the publication. An OPF Package Document must include at least one instance of this element type, however `multiple instances are permitted`. Any Reading System that displays title metadata should display the content of `the most appropriate title element(s)`. Determination of the most appropriate titles is not defined by this specification, but may include the available fonts, an examination of xml:lang attributes, or other heuristics. In the absence of such an algorithm, conforming Reading Systems should consider either the first title element or all the title elements as the most appropriate._ 

I guess this means there can be multiple <title>, one in plain text, one in HTML, and we are safe to use only the first as we do now.

#### subject:
_Multiple instances of the subject element are supported, each including an arbitrary phrase or keyword. This specification makes no attempt to standardize subject naming schemes, such as the Library of Congress Subject Heading System._ 

We already deal with that: concatenate them, separated by `; `.

#### language:
_Identifies a language of the intellectual content of the Publication. An OPF Package Document must include at least one instance of this element type, however `multiple instances are permitted`. The content of this element must comply with RFC 3066_...

Setting multiple lang in one variable (`fr; en; de`) may interfere or needs adaptation of stuff that currently use it (choice of hyphenation dict, wikipedia choice of default langages, may be choice of OCR lang but that shouldn't matter for epub).

**Comment by Frenzie** [comment]
> If we should concatenate the multiple <dc:creator> into our single authors field, what's should the separator be? ; like for keywords?

For authors `&` could also make sense.

**Comment by poire-z** [comment]
No problem when more than 2? 
`Lennon, John & Marley, Bob & Hendrix, Jimmy & Mamas & the papas, The`
(a `<dc:creator>` content may include a `&` - I guess it's less likely it includes a `;`)
(I also guess bands don't write books :) but that was just for example)

**Comment by Frenzie** [comment]
It's not ideal obviously but I wouldn't call it a problem as such. It kind of feels more natural than `;` to me but I'm sure it's fine either way.

**Comment by poire-z** [comment]
Would it look too ugly if, in crengine, we used a tabulation (or some other control ascii code that we're sure to not find in `<dc:creator>` XML) to join the multiple `<dc:creator>` contents into a single string?
We could then do the final `tab` to `&` conversion in lua in `frontend/document/credocument.lua` and it would be easier to fix/tweak, or use some setting to use `&`, `;`, or to just use the first item.

**Comment by Frenzie** [comment]
From a crengine POV that sounds terrible; from our POV that's probably okay? Like I said though I have no objections to `;`.

**Comment by poire-z** [comment]
I'm mostly thinking about a case where 20 contributors to a publication would be listed in the authors tags (i guess I have no such books, I don't read much collective articles and such).
If some user is bothered by that, I'd rather like to have to fix that on the lua side (with a `G_reader_setting:isTrue("crengine_ignore_multiple_authors")` ) than on the crengine C side :)

**Comment by Frenzie** [comment]
Oh, it's very simple to come up with a book that contains the work of many authors. Random pick: http://www.oapen.org/search?identifier=641876

However, finding an ebook that actually puts that information:

a. in the metadata
b. is an EPUB

*That* I have no opinion on. I think https://www.doabooks.org/ is also completely PDF.

On my Humble Bundle I see *2013 Nebula Awards Showcase*, which in the PDF metadata — I normally use the PDFs — merely says "Edited by Catherine Asaro"

The EPUB does effectively the same thing:

```xml
  <metadata xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:opf="http://www.idpf.org/2007/opf">
    <dc:identifier id="BookId" opf:scheme="UUID">urn:uuid:39c74496-f15e-4424-b1b4-3da08bdd8a0f</dc:identifier>
    <dc:title>Nebula Awards Showcase 2013</dc:title>
    <dc:creator opf:role="aut">Catherine Asaro</dc:creator>
    <dc:language>en</dc:language>
    <dc:identifier opf:scheme="ISBN">9781616147846</dc:identifier>
    <dc:publisher>Prometheus Books</dc:publisher>
    <dc:date opf:event="modification">2014-04-15</dc:date>
    <meta content="0.7.4" name="Sigil version" />
    <meta name="cover" content="cover.jpg" />
  </metadata>
```

tl;dr Finding multi-author PDFs: piece of cake; finding multi-author EPUBs not too hard either. Finding 'em with such metadata, on the other hand… honestly, I'm surprised the metadata on these is as good as it is. Much better than usual.

**Comment by poire-z** [comment]
I found 2 epub with more than one `<dc:creator>` among my emulator test files.

One has the preface author and a foreword author (too much information for me).
One even has the author's firstname and lastname in separate tags...:
```xml
    <dc:creator opf:file-as="Schiffter" opf:role="aut">Schiffter</dc:creator>
    <dc:creator opf:file-as="Frédéric" opf:role="aut">Frédéric</dc:creator>
```
How it would look with `&` as the separator:

<kbd>![list_ampersand](https://user-images.githubusercontent.com/24273478/35159137-893c8152-fd39-11e7-89d8-e2c8d2a6ce91.png)</kbd>

<kbd>![mosaic_ampersand](https://user-images.githubusercontent.com/24273478/35159141-8c1b63de-fd39-11e7-958e-6018dad4d3a2.png)</kbd>

With ` ; ` as the separator (with space before and after - should there be no space before?):

<kbd>![list_semicolon](https://user-images.githubusercontent.com/24273478/35159253-e79d2012-fd39-11e7-83e0-19c1f184944b.png)</kbd>

<kbd>![mosaic_semicolon](https://user-images.githubusercontent.com/24273478/35159260-ea036348-fd39-11e7-89d0-2190fca94699.png)</kbd>

(With ` ; ` the bogus lastname-firstname-authors looks less strange).

Would need that tabulation thing in crengine, and this added to credocument.lua:
```lua
function CreDocument:getProps()
    local props = self._document:getDocumentProps()
    local authors = props.authors
    if authors:find("\t") then
        -- crengine may give us multiple authors (when there are multiple
        -- <dc:creator> tags), separated in this string with tab
        authors = util.splitToArray(authors, "\t")
        if G_reader_setting.isTrue("crengine_ignore_multiple_authors") then
            props.authors = table.remove(authors, 1) -- use only the first one
        else
            props.authors = table.concat(authors, " & ") -- or "; "
        end
    end
    return props
end
```
Should I go on with that?
With & or ; ?
How to name the manually add'able setting and what should be the default?

**Comment by gerhaher** [comment]
I cast my vote for '&'.

**Comment by Frenzie** [comment]
Like I already said I think `&` is the best simple solution. The fancier solution is `,` and `&` as the last one.

**Comment by poire-z** [comment]
The fancier solution is do'able (as I don't have to do it in C :). But a bit confusing when individual author items are of the form "Lastname, Firstname".
Dunno if that happens often, in @gerhaher example, there is: `<dc:creator xmlns:ns1="http://www.idpf.org/2007/opf" ns1:role="aut" ns1:file-as="Berglund, Tobias & Sennerteg, Niclas">Tobias Berglund</dc:creator>` so the content may be the human friendly "Firstname Lastname", and the file-as contains the good-for-alphabetical sorting "Lastname, Firstname".

(Some answers to my other questions? default, setting name :)

**Comment by Frenzie** [comment]
I mildly dislike the fancier solution because it only really works properly for

Surname1, Firstname1, Firstname2 Surname2, Firstname3 Surname3 and Firstname4 Surname 4.

I would probably just show all the names with `&`.

**Comment by poire-z** [comment]
Just thought I'd be using `\n` instead of `\t` on the crengine side for safety (other book formats may have \t in their meta, or even \n).

And it naturally gave:

<kbd>![list_cr](https://user-images.githubusercontent.com/24273478/35170198-9feab538-fd5f-11e7-958a-4ff4b3987fed.png)</kbd>

<kbd>![mosaic_cr](https://user-images.githubusercontent.com/24273478/35170285-f6c38952-fd5f-11e7-90e9-dafa57f9b704.png)</kbd>

which look quite OK when only 2-3 authors.

There are squares (undisplayable char) in the Book info KeyValuePage, but when you Hold on it to open textviewer, you get this nice _one author per line_ display.

**Comment by Frenzie** [comment]
And of course when there are more than *n* authors you say et alia. Where *n* is something from 3 to 5.

**Comment by poire-z** [comment]
Would look neat indeed :)
I'm more used to see the shorter _et al._
Should this be on a line of its own or appened to the last kep author?
Guess this neat treatment would have to be done in coverbrowser render code, so we keep all the authors in book db and we see them all in the Book information KeyValuePage.

**Comment by poire-z** [comment]
Proposed solution in #3624.
